x-pack/plugins/monitoring/server/alerts/memory_usage_alert.ts

```diff
@@ -0,0 +1,355 @@
+        state: AlertingDefaults.ALERT_STATE.firing,
+        nodes: firingNodes
+          .map((state) => `${state.nodeName}:${state.memoryUsage.toFixed(2)}`)
+          .join(','),
+        count: firingCount,
+        clusterName: cluster.clusterName,
+        action,
+        actionPlain: shortActionText,
+      });
+    } else {
+      const resolvedNodes = (alertStates as AlertMemoryUsageState[])
+        .filter((state) => !state.ui.isFiring)
+        .map((state) => `${state.nodeName}:${state.memoryUsage.toFixed(2)}`);
+      const resolvedCount = resolvedNodes.length;
+
+      if (resolvedCount > 0) {
+        const internalMessage = i18n.translate(
+          'xpack.monitoring.alerts.memoryUsage.resolved.internalMessage',
+          {
+            defaultMessage: `Memory usage alert is resolved for {count} node(s) in cluster: {clusterName}.`,
+            values: {
+              count: resolvedCount,
+              clusterName: cluster.clusterName,
+            },
+          }
+        );
+
+        instance.scheduleActions('default', {
+          internalShortMessage: internalMessage,
+          internalFullMessage: internalMessage,
+          state: AlertingDefaults.ALERT_STATE.resolved,
+          nodes: resolvedNodes.join(','),
+          count: resolvedCount,
+          clusterName: cluster.clusterName,
+        });
+      }
+    }
+  }
+
+  protected async processData(
+    data: AlertData[],
+    clusters: AlertCluster[],
+    services: AlertServices,
+    logger: Logger,
+    state: any
+  ) {
+    const currentUTC = +new Date();
+    for (const cluster of clusters) {
+      const nodes = data.filter((node) => node.clusterUuid === cluster.clusterUuid);
+      if (!nodes.length) {
+        continue;
+      }
+
+      const firingNodeUuids = nodes
+        .filter((node) => node.shouldFire)
+        .map((node) => node.meta.nodeId)
+        .join(',');
+      const instanceId = `${this.type}:${cluster.clusterUuid}:${firingNodeUuids}`;
+      const instance = services.alertInstanceFactory(instanceId);
+      const newAlertStates: AlertMemoryUsageState[] = [];
+
+      for (const node of nodes) {
+        const stat = node.meta as AlertMemoryUsageState;
+        const nodeState = this.getDefaultAlertState(cluster, node) as AlertMemoryUsageState;
+        nodeState.memoryUsage = stat.memoryUsage;
+        nodeState.nodeId = stat.nodeId;
+        nodeState.nodeName = stat.nodeName;
+
+        if (node.shouldFire) {
+          nodeState.ui.triggeredMS = currentUTC;
+          nodeState.ui.isFiring = true;
+          nodeState.ui.severity = node.severity;
+          newAlertStates.push(nodeState);
+        }
+        nodeState.ui.message = this.getUiMessage(nodeState, node);
+      }
+
+      const alertInstanceState = { alertStates: newAlertStates };
+      instance.replaceState(alertInstanceState);
+      if (newAlertStates.length && !instance.hasScheduledActions()) {
```

Comment:
That's true, however, this is to avoid any race conditions for if they set the check time to something really small (I was able to confirm this), so just as an extra precaution I don't think it would hurt

Reply:
@mikecote Do you have any insight into if we should be calling this? Does it not matter?

Reply:
The function `hasScheduledActions()` is the same as asking "did I call `scheduleActions`?" It's mostly designed to be used internally to see if there's any actions to schedule for each alert instance. The throttle doesn't impact the result of this function.

You can see the internals of that function [here](https://github.com/elastic/kibana/blob/master/x-pack/plugins/alerts/server/alert_instance/alert_instance.ts#L35) and how the variable is only set [here](https://github.com/elastic/kibana/blob/master/x-pack/plugins/alerts/server/alert_instance/alert_instance.ts#L71-L75).

Reply:
@igoristic It seems safe by the way they are using it. It feels unnecessary to me, but whichever is fine with me

Reply:
The only reason I added it was because I was hitting: https://github.com/elastic/kibana/blob/master/x-pack/plugins/alerts/server/alert_instance/alert_instance.ts#L69 with really small check times, but I don't think this would happen in a real use cases anyways. Removed it for now until proven otherwise